commands/volumes/volume-create.go

```diff
@@ -43,68 +47,117 @@ func createVolume(msg *volume.VolCreateRequest) (*volume.Volinfo, error) {
 	if err != nil {
-		return httpStatusCode, err
+		return err
 	}
-	return 0, nil
+
+	_, err = volume.ValidateBrickEntriesFunc(vol.Bricks, vol.ID, req.Force)
+	if err != nil {
+		return err
+	}
+
+	// Store volinfo for later usage
+	c.Set("volinfo", vol)
+
+	return nil
 }
 
-func commitVolumeCreate(vol *volume.Volinfo) (int, error) {
+func generateVolfiles(c *context.Context) error {
+	vol, ok := c.Get("volinfo").(*volume.Volinfo)
+	if !ok {
+		return errors.New("failed to get volinfo from context")
+	}
+
 	// Creating client and server volfile
 	e := volgen.GenerateVolfileFunc(vol)
 	if e != nil {
-		log.WithFields(log.Fields{"error": e.Error(),
+		c.Log.WithFields(log.Fields{"error": e.Error(),
 			"volume": vol.Name,
-		}).Error("Failed to generate volfile")
-		return http.StatusInternalServerError, e
+		}).Error("failed to generate volfile")
+		return e
 	}
+	return nil
+}
 
-	e = volume.AddOrUpdateVolumeFunc(vol)
+func storeVolume(c *context.Context) error {
+	vol, ok := c.Get("volinfo").(*volume.Volinfo)
+	if !ok {
+		return errors.New("failed to get volinfo from context")
+	}
+
+	e := volume.AddOrUpdateVolumeFunc(vol)
 	if e != nil {
-		log.WithFields(log.Fields{"error": e.Error(),
+		c.Log.WithFields(log.Fields{"error": e.Error(),
 			"volume": vol.Name,
 		}).Error("Failed to create volume")
-		return http.StatusInternalServerError, e
+		return e
 	}
-	log.WithField("volume", vol.Name).Debug("NewVolume added to store")
-	return 0, nil
+
+	log.WithField("volume", vol.Name).Debug("new volume added")
+	return nil
 }
 
-func rollBackVolumeCreate(vol *volume.Volinfo) error {
+func rollBackVolumeCreate(c *context.Context) error {
+	vol, ok := c.Get("volinfo").(*volume.Volinfo)
+	if !ok {
+		return errors.New("failed to get volinfo from context")
+	}
+
 	_ = volume.RemoveBrickPaths(vol.Bricks)
+
 	return nil
 }
 
 func volumeCreateHandler(w http.ResponseWriter, r *http.Request) {
+	req := new(volume.VolCreateRequest)
 
-	msg := new(volume.VolCreateRequest)
-
-	httpStatusCode, e := validateVolumeCreateJSONRequest(msg, r)
+	httpStatus, e := unmarshalVolCreateRequest(req, r)
 	if e != nil {
-		rest.SendHTTPError(w, httpStatusCode, e.Error())
+		rest.SendHTTPError(w, httpStatus, e.Error())
 		return
 	}
-	vol, e := createVolume(msg)
-	if e != nil {
-		rest.SendHTTPError(w, http.StatusInternalServerError, e.Error())
-		return
+
+	// TODO: Properly construct these things
+	nodes := make([]string, 1)
+	c := context.NewLoggingContext(log.Fields{
+		"reqid": uuid.NewRandom().String(),
+	})
+	c.Set("req", req)
+
+	txn := &transaction.SimpleTxn{
+		Ctx:      c,
+		Nodes:    nodes,
+		LockKey:  req.Name,
+		Stage:    validateVolumeCreate,
+		Commit:   generateVolfiles,
+		Store:    storeVolume,
+		Rollback: rollBackVolumeCreate,
 	}
-	httpStatusCode, e = validateVolumeCreate(msg, vol)
-	if e != nil {
-		rest.SendHTTPError(w, httpStatusCode, e.Error())
+
+	if e := txn.Do(); e != nil {
+		rest.SendHTTPError(w, http.StatusInternalServerError, e.Error())
```

Comment:
So this indicates we are reporting back a generic http status code in case of an error. Earlier we were using different status codes. Is it something we can't maintain the same?